src/Lib.P2P/Transports/Tcp.cs

```diff
@@ -0,0 +1,293 @@
+                }
+            }
+
+            _ = Task.Run(() => ProcessConnection(socket, address, handler, cancel), cancel);
+
+            return address;
+        }
+
+        private static void ProcessConnection(Socket socket,
+            MultiAddress address,
+            Action<Stream, MultiAddress, MultiAddress> handler,
+            CancellationToken cancel)
+        {
+            Log.Debug("listening on " + address);
+
+            // Handle cancellation of the listener
+            cancel.Register(() =>
+            {
+                Log.Debug("Got cancel on " + address);
+
+                try
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                    {
+                        socket.Shutdown(SocketShutdown.Both);
+                        socket.Dispose();
+                    }
+                    else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                    {
+                        socket.Shutdown(SocketShutdown.Receive);
+                    }
+                    else // must be windows
+                    {
+                        socket.Dispose();
+                    }
+                }
+                catch (Exception e)
+                {
+                    Log.Warn($"Cancelling listener: {e.Message}");
+                }
+                finally
+                {
+                    socket = null;
+                }
+            });
+
+            try
+            {
+                while (!cancel.IsCancellationRequested)
+                {
+                    var conn = socket.Accept();
+
+                    MultiAddress remote = null;
+                    if (conn.RemoteEndPoint is IPEndPoint endPoint)
+                    {
+                        var s = new StringBuilder();
+                        s.Append(endPoint.AddressFamily == AddressFamily.InterNetwork ? "/ip4/" : "/ip6/");
+                        s.Append(endPoint.Address);
+                        s.Append("/tcp/");
+                        s.Append(endPoint.Port);
+                        remote = new MultiAddress(s.ToString());
+                        Log.Debug("connection from " + remote);
+                    }
+
+                    conn.NoDelay = true;
+                    Stream peer = new NetworkStream(conn, true);
+
+                    peer = new DuplexBufferedStream(peer);
+
+                    try
+                    {
+                        handler(peer, address, remote);
+                    }
+                    catch (Exception e)
+                    {
+                        Log.Error("listener handler failed " + address, e);
+                        peer.Dispose();
+                    }
+                }
+            }
+            catch (Exception) when (cancel.IsCancellationRequested)
+            {
+                // ignore
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.Message);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this logging statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)